### ISSUE TITLE
Confusion matrix calculation in val.py

### ISSUE THREAD

**Issue by Nomow** [ultralytics/yolov5#4756]
I'm a bit confused about the confusion matrix calculation in **val.py**.

```
            if len(pred) == 0:
                if nl:
                    stats.append((torch.zeros(0, niou, dtype=torch.bool), torch.Tensor(), torch.Tensor(), tcls))
                continue

```
if there are no predictions, but there are ground truth data, shouldn't we incrimate confusion_matrix? Wouldn't it count as a FN? 
Because now it seems we only incremate the confusion matrix when we have both ground_truth and detections.

**Comment by glenn-jocher** [comment]
@Nomow yes, this is a good point! In theory this may appear to be a problem, but in practice with default `conf_thres=0.001` this `if` statement you have shown has an extremely low likelihood of ever passing `True`.

Nevertheless if you would like to propose a fix please submit a PR. Please see our ✅ [Contributing Guide](https://docs.ultralytics.com/help/contributing/) to get started. Thanks!

**Comment by Nomow** [comment]
@glenn-jocher Thank you for the response.
I will make a PR, but just to make sure, before I make one:

```
            # Evaluate
            if nl:
                tbox = xywh2xyxy(labels[:, 1:5])  # target boxes
                scale_coords(img[si].shape[1:], tbox, shape, shapes[si][1])  # native-space labels
                labelsn = torch.cat((labels[:, 0:1], tbox), 1)  # native-space labels
                correct = process_batch(predn, labelsn, iouv)
                if plots:
                    confusion_matrix.process_batch(predn, labelsn)
            else:
                #### if there are no ground truth, but are predictions then increment FP
                correct = torch.zeros(pred.shape[0], niou, dtype=torch.bool)
            stats.append((correct.cpu(), pred[:, 4].cpu(), pred[:, 5].cpu(), tcls))  # (correct, conf, pcls, tcls)

```

In the if-else statement, if there is no ground truth data, and there are detections within the image, then i think we should increment CM as well. Technically it should be an FP detection, or is there any reason why we don't increment it? Or its because there are rare cases when there are no labels for image?

**Comment by glenn-jocher** [comment]
@Nomow the fastest and easiest way to incorporate your ideas into the official codebase is to submit a Pull Request (PR) implementing your idea, and if applicable providing before and after profiling/inference/training results to help us understand the improvement your feature provides. This allows us to directly see the changes in the code and to understand how they affect workflows and performance.
